src/api/Base.js

```diff
@@ -165,54 +193,152 @@ class Base {
-        this.errorCallback = errorCallback;
+        this.attachCallbacks(successCallback, errorCallback);
 
         // Make the XHR request
         try {
-            const { data } = await this.getData(id, params);
+            const { data } = await this.xhr.get({
+                id: getTypedFileId(id),
+                url: this.getUrl(id),
+                params
+            });
+
             this.successHandler(data);
         } catch (error) {
             this.errorHandler(error);
         }
     }
+
+    /**
+     * Generic API post
+     *
+     * @param {string} id - The file id
+     * @param {string} url - The url to post to
+     * @param {Object} data - The data to post
+     * @param {Function} successCallback - The success callback
+     * @param {Function} errorCallback - The error callback
+     */
+    async post(
+        id: string,
+        url: string,
+        data: Object,
+        successCallback: Function,
+        errorCallback: Function
+    ): Promise<void> {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.attachCallbacks(successCallback, errorCallback);
+
+        // Make the XHR request
+        try {
+            const { responseData } = await this.xhr.post({ id, url, data });
+            this.successHandler(responseData);
+        } catch (error) {
+            this.errorHandler(error);
+        }
+    }
+
+    /**
+     * Generic API put
+     *
+     * @param {string} id - The file id
+     * @param {string} url - The url to post to
+     * @param {Object} data - The data to post
+     * @param {Function} successCallback - The success callback
+     * @param {Function} errorCallback - The error callback
+     */
+    async put(
+        id: string,
+        url: string,
+        data: Object,
+        successCallback: Function,
+        errorCallback: Function
+    ): Promise<void> {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.attachCallbacks(successCallback, errorCallback);
+
+        // Make the XHR request
+        try {
+            const { responseData } = await this.xhr.put({ id, url, data });
+            this.successHandler(responseData);
+        } catch (error) {
+            this.errorHandler(error);
+        }
+    }
+
+    /**
+     * Generic API post
+     *
+     * @param {string} id - The file id
+     * @param {string} url - The url to post to
+     * @param {Function} successCallback - The success callback
+     * @param {Function} errorCallback - The error callback
+     * @param {Object} data optional data to delete
+     */
+    async delete(
+        id: string,
+        url: string,
+        successCallback: Function,
+        errorCallback: Function,
+        data?: Object = {}
+    ): Promise<void> {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.attachCallbacks(successCallback, errorCallback);
+
+        // Make the XHR request
```

Comment:
It looks like all these non-get functions are same, only thing thats diff is this.xhr[operation]